### ISSUE TITLE
Upgrade jobs to work with Salesforce adaptor `v4.6.5`

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/women-for-women#159]
## Background, context, and business value
Once we upgraded the SF adaptor version from `v2.7.3` to `v4.6.5` so that we can use Oauth, we're now seeing job run errors like this: `{"message":"TypeError: Cannot read properties of undefined (reading 'includes')","name":"RuntimeError"}`

Examples: 
- [Upsert Donors run failure](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/runs/a99f2e8e-1562-4e6e-97d9-60f02fc0f2a3?step=9ff57c77-065a-4df9-8ee4-9f53ef9b1f53)
- [Direct Debits run failure](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/runs/08238f63-9624-41d2-9853-00f7180c67eb?step=2d5f181b-4d63-43a3-a380-bc74e06617d9)

From Mtuchi: there might be an issue with how the Query returns when no records returned. 

## The specific request, in as few words as possible
Please update the following 5 jobs to work with the latest SF adaptor version. 

## state.json
NOTE: We use SF Oauth on app.openfn.org (see LP `API user - WFW Salesforce Sandbox Permissions (Q2 2024)`) 

For local testing, you can use the Admin user for basic auth: 
```
{
  "configuration": ["SEE LAST PASS: 'Admin user - WFW Salesforce Permissions Sandbox (Q2 2024)'"],
}
```

## expression.js
1. [upsertDonorsV2.js](https://github.com/OpenFn/women-for-women/blob/master/jobs/committed/upsertDonorsV2.js)
2. [upsertDirectDebits.js](https://github.com/OpenFn/women-for-women/blob/master/jobs/committed/upsertDirectDebits.js)
3. [upsertTransactionDDV2.js](https://github.com/OpenFn/women-for-women/blob/master/jobs/committed/upsertTransactionDDV2.js)
4. [upsertCardMasterV2.js](https://github.com/OpenFn/women-for-women/blob/master/jobs/committed/upsertCardMasterV2.js)
5. [upsertTransactionCardsV2.js](https://github.com/OpenFn/women-for-women/blob/master/jobs/committed/upsertTransactionCardsV2.js)

## To test
You can run these 3 workflows on app to test for `Donors`, `Debit`, and `Credit`
https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w

`manualCursor` to use when testing - try `"20240501"` and `"20240425"`

**Comment by mtuchi** [comment]
I have made improvements to job codes to be able to support latest version of salesforce. 
For donors workflow #154 , For Credit & Debit see #153

**Comment by mtuchi** [comment]
@aleksa-krolls i am getting duplicates error for donors workflow, [See run log](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w/7a2aacd9-0f47-4e1f-b89d-e3ceca028d04?a=826b798f-49a4-4be8-bc70-9a3abb045711&m=expand&s=38abc66d-59d0-4547-84da-02511169b6cf). And for debit workflow i am getting [Foreign key external ID:  not found ](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w/46758111-9ef9-483b-be63-386b51c11107?a=79d4cf36-5bf8-4af6-a900-64800ba13ab4&m=expand&s=5562a09e-99eb-4c47-8e23-7f35b63b559d)

**Comment by aleksa-krolls** [comment]
@mtuchi to unblock you here...
First, I fixed the donors dupe error, but now the job is failing with a different error - [see run failure](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w/7a2aacd9-0f47-4e1f-b89d-e3ceca028d04?a=c6dd5b29-34bd-4ea7-8f5c-4c323226d3bb&m=expand&s=38abc66d-59d0-4547-84da-02511169b6cf). I think this error is related to when something is wrong with the external Id... so I worry we're not mapping `wfw_Legacy_Supporter_ID__c` properly. Can you pls take a look and ping me on Slack to discuss/if different errors pop up? 
```
Upserting Contact with externalId wfw_Legacy_Supporter_ID__c 
Failed step Upsert Donors after 13.39s
AdaptorError: HTTP Method 'PATCH' not allowed. Allowed are GET,HEAD,POST
{"message":"HTTP Method 'PATCH' not allowed. Allowed are GET,HEAD,POST","name":"AdaptorError"}
```
Second, regarding the [failed debit workflow run](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w/46758111-9ef9-483b-be63-386b51c11107?a=79d4cf36-5bf8-4af6-a900-64800ba13ab4&m=expand&s=5562a09e-99eb-4c47-8e23-7f35b63b559d) you shared - it looks like you're trying to sync debit data from the csv file `20240501`, is that right? If yes, please make sure all donors are synced from the `wfwi donors 20240501` file _before_ running this debit workflow. Pls try that and then ping me on Slack if you're still running into errors.

**Comment by mtuchi** [comment]
@aleksa-krolls the mapping for `wfw_Legacy_Supporter_ID__c` is `wfw_Legacy_Supporter_ID__c: x.PersonRef && x.PersonRef !== '' ? x.PersonRef : undefined,`. I have looked at the record that is causing the error and the value for `"PersonRef": "",` which will make the value for `wfw_Legacy_Supporter_ID__c: undefined`. 
Can you advice how we should update the mapping ?

**Comment by aleksa-krolls** [comment]
@mtuchi I don't think it's a mapping issue... rather I think the logic has maybe been mistakenly edited. [See this diagram](https://lucid.app/lucidchart/4d0b44a8-3299-431f-84b5-4b00bc713cd7/edit?page=JtuY-xeSZoyq&invitationId=inv_42d970ff-eecc-437c-95a0-19688f2b2d6e#) for the donors sync workflow. 

In query 2, if no Contact can be found with a SF `wfw_Legacy_Supporter_ID__c` that matches the CG `PersonRef`, then the job should move on to query 3....
<img width="373" alt="Screenshot 2024-05-30 at 1 01 14 AM" src="https://github.com/OpenFn/women-for-women/assets/16758106/d8dc642d-46eb-4b71-bc7f-00db0dc4127f">

However, per [the run log](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w/7a2aacd9-0f47-4e1f-b89d-e3ceca028d04?a=c6dd5b29-34bd-4ea7-8f5c-4c323226d3bb&m=expand&s=38abc66d-59d0-4547-84da-02511169b6cf), it looks like you're trying to upsert the Contact even though no externalId is provided... but I think rather the job should move on to query 3.  
<img width="926" alt="Screenshot 2024-05-30 at 1 05 00 AM" src="https://github.com/OpenFn/women-for-women/assets/16758106/853838e3-e101-4e36-807c-b4b7a57dc7dc">

**Comment by mtuchi** [comment]
@aleksa-krolls i found the problem that was causing this behavior, It a small regression that i introduced in our salesforce adaptor. I have a fix for it here https://github.com/OpenFn/adaptors/pull/590, I will let you know once the new release is available so that you can re-test the workflow again

**Comment by aleksa-krolls** [comment]
@mtuchi when will you be ready to work on this? And do you need to pair with someone to finish this? We need to finish this before you go offline this week. Also please update the estimate or create a new issue for any work that needs to be done for this sprint... adding to top of Backlog for now and hoping you can work on it soon.

**Comment by daissatou2** [comment]
Hey @aleksa-krolls what's the latest here? Is this project ready for V2 training?

**Comment by mtuchi** [comment]
Hey @daissatou2 this PR was waiting a new salesforce adaptor [v4.8.0](https://github.com/OpenFn/adaptors/pull/632/files#diff-720a7ece01a94dfc91055e4adb2b05661225385afacd38af31b2e510b62f687b) which was released when i was offline. My bad 🙌🏽  i should have share EOD updates before i went offline. 
cc @aleksa-krolls

**Comment by mtuchi** [comment]
@Aleksa Krolls i have updated salesforce adaptor version in women-for-women

[Donors, Debit & Credit workflow](https://app.openfn.org/projects/d84ec7c9-7b5a-4f01-a0d2-88f359deb0e2/w), I have re-run the Donors workflow for `wfwi Donors 20240501.csv`. But i am getting a salesforce error saying 

```
duplicate value found: wfw_Legacy_Supporter_ID__c duplicates value on record with id: 003O300000Eh8DrIAJ, Committed_Giving_ID__c duplicates value on record with id: 003O300000Eh8DrIAJ
```
cc @daissatou2

**Comment by mtuchi** [comment]
@daissatou2 what is the status here ?

**Comment by daissatou2** [comment]
@mtuchi in Aleksa's court
